574 table support column alignment

sunilkotiyaibm opened the pull request:
<!-- Please fill in areas below: -->

**Summary**

Introduced a non-mandatory align property in table columns to align table content including header, align property value could be start, center or end.

**Change List (commits, features, bugs, etc)**

src/components/DataTable/_data-table.scss
src/components/Table/Table.story.jsx
src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx
src/components/Table/TableHead/TableHead.jsx
src/components/Table/TablePropTypes.js
src/components/__snapshots__/StorybookSnapshots.test.js.snap

**Related Issues**

<!-- replace NUMBER with issue number to auto-close on merge -->

- Closes IBM/carbon-addons-iot-react#574

**Acceptance Test (how to verify the PR)**

- Change the table align property (start, center or end) in Table.story.jsx for const tableColumn,  changes will reflect the respective table column in storybook-Table.
e.g. const tableColumns = [
  {
    id: 'string',
    name: 'String',
    filter: { placeholderText: 'pick a string' },
    align: 'start',
  }

stuckless commented on src/components/Table/TablePropTypes.js:
```diff
@@ -82,6 +82,7 @@ export const TableColumnsPropTypes = PropTypes.arrayOf(
     name: PropTypes.string.isRequired,
     isSortable: PropTypes.bool,
     width: PropTypes.string, // ex: 150px, or 2rem
+    align: PropTypes.string, // ex: start, center, end
```
This should be an enum with 3 options 'start', 'center' and 'end'.

Also we should set the default props on the column to be start so that we don't need to explicitly say 'start' when defining a column.

sunilkotiyaibm replied:
Thanks Sean, we will incorporate your suggestion.

sunilkotiyaibm replied:
Hi Sean, Code is updated based on your review comments, please review.

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,62 @@
+html[dir='rtl'] {
+  td.data-table {
+    &-start {
+      padding-right: 1.25rem;
```
Spacing values like padding, margin, etc. should not be hardcoded, but instead use [sass variables provided by carbon](https://www.carbondesignsystem.com/guidelines/spacing#spacing-scale)

sunilkotiyaibm replied:
Now we are able to managed using saas variable i.e.  padding-right: $spacing-09 + $spacing-02;  for 3.25rem

sunilkotiyaibm replied:
Code is updated based on your review comments, please review.

tay1orjones commented on src/components/Table/TableHead/TableHead.jsx:
```diff
@@ -178,6 +178,7 @@ const TableHead = ({
                 }
               }}
               sortDirection={hasSort ? sort.direction : 'NONE'}
+              className={`bx--table-header-label-${matchingColumnMeta.align}`}
```
Class names should not use a hardcoded `bx` prefix as this namespace is configurable by the end user via `$prefix` - [Link to carbon documentation](https://www.carbondesignsystem.com/get-started/develop/vanilla#namespacing)

sunilkotiyaibm replied:
Code is updated based on your review comments, Please review.

davidicus commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,62 @@
+
+.bx--table-header-label {
+  &-start {
+    span.bx--table-header-label {
```
if not necessary for specificity we should use the class here without the element selector.

sunilkotiyaibm replied:
Code is updated based on your review comments (removed span), Please review.

davidicus commented on src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx:
```diff
@@ -322,17 +322,17 @@ const StyledExpansionTableRow = styled(({ hasRowSelection, ...props }) => <Table
+        text-align: ${align};
       `
-        : `padding-right: ${props.offset}px`;
+        : `text-align:${props.align}`;
```
should be able to use `align` instead of props.align

sunilkotiyaibm replied:
Code is updated based on your review comments, Please review.

davidicus commented on src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx:
```diff
@@ -419,6 +419,8 @@ const TableBodyRow = ({
             offset={offset}
             width={matchingColumnMeta && matchingColumnMeta.width}
+            align={matchingColumnMeta && matchingColumnMeta.align}
+            className={col.isSortable === true ? `data-table-${matchingColumnMeta.align}` : ``}
```
is there a reason we only do this if the row is sortable?

sunilkotiyaibm replied:
Alignment class added in both case 1. With sorting icon 2. Without sorting icon

sunilkotiyaibm replied:
Code is updated based on your review comments, Please review.

stuckless commented:
@sunilkotiyaibm any followup on the items in the code review?

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+}
+
+.bx--data-table td:last-of-type {
+  padding-right: $spacing-09 + $spacing-02;
```
Is there a design spec to show why we're deviating from the Carbon values for padding in table cells? I suggest we use the standard values to be in line with Carbon: https://github.com/carbon-design-system/carbon/blob/c87a493f8dd86b5289fda9e59a4ee84639652028/packages/components/src/components/data-table/_data-table-core.scss#L88-L94

sunilkotiyaibm replied:
We are not able to align table column header and its respective data in cells while using standard spacing. i.e for 3.25 rem for which we used $spacing-09 + $spacing-02, but if we use $spacing-09 only then it is not aligning as showing in attached screenshot shared over slack

davidicus replied:
I think we need to just use the `text-align` property here. The padding of these cells should use the standard padding we get from the Carbon component.

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+  }
+}
+
+.bx--data-table td:last-of-type {
```
`bx` needs to be replaced with the prefix variable

sunilkotiyaibm replied:
Ok replaced with #{$prefix}

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+
+.table-header-label {
+  &-start {
+    .bx--table-header-label {
```
prefix variable instead of `bx`

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+  }
+
+  &-center {
+    .bx--table-header-label {
```
prefix variable instead of `bx`

sunilkotiyaibm replied:
Ok replaced with #{$prefix}

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+  }
+
+  &-end {
+    .bx--table-header-label {
```
prefix variable instead of `bx`

sunilkotiyaibm replied:
Ok replaced with #{$prefix}

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+  padding-right: $spacing-09 + $spacing-02;
+}
+
+.bx--data-table td:first-of-type {
```
prefix variable instead of `bx`

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+  padding-left: $spacing-04 + $spacing-02;
+}
+
+.bx--table-sort__icon-unsorted,
```
prefix variable instead of `bx`

sunilkotiyaibm replied:
Ok replaced with #{$prefix}

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+    padding-right: $spacing-09 + $spacing-02;
+  }
+
+  padding-right: $spacing-04 + $spacing-02 !important;
```
please justify the use of `!important` - this shouldn't be necessary to apply the required styles atop Carbon's styling

davidicus replied:
This has been removed.

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+      padding-right: $spacing-05 + $spacing-02;
+    }
+
+    padding-right: $spacing-04 + $spacing-02 !important;
```
please justify the use of `!important` - this shouldn't be necessary to apply the required styles atop Carbon's styling

davidicus replied:
I have refactored all the padding changes out of this PR

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+
+  &-end {
+    .bx--table-header-label {
+      width: 100%;
```
This declaration is common among this entire block, de-dupe by placing in a selector not scoped by `&-center` or `&-end`, etc

davidicus replied:
done

tay1orjones commented on src/components/DataTable/_data-table.scss:
```diff
@@ -1 +1,67 @@
+  }
+
+  &-end {
+    padding-left: $spacing-05 + $spacing-02;
```
This declaration is common among `&-start`, `&-end`, and the rtl `&-start` below. De dupe this by placing in a unified selector. It seems that all table cells with `start` should have this left padding.

davidicus replied:
done.

tay1orjones commented on src/components/Table/Table.story.jsx:
```diff
@@ -66,30 +66,36 @@ export const tableColumns = [
     id: 'string',
     name: 'String',
     filter: { placeholderText: 'pick a string' },
+    align: 'start',
```
Please create a separate story to showcase this functionality and scope the feature to only that story instead of globally for all stories. Ideally this parameter would also be a knob so users can play with the different states of `start`, `end`, and `center`

davidicus replied:
done.

davidicus replied:
Did not incorporate a knob though

tay1orjones commented:
@JordanWSmith15 @KarenGski @lukefirth A question for design - based on this spec, there is a question regarding the right alignment present in this PR.
![image](https://user-images.githubusercontent.com/3360588/67951892-621bf900-fbba-11e9-8882-9934cac3e0ba.png)

This shows right alignment should take into account the sorting icon and align content with the label, but this seems nonstandard to me.

When left and center aligned, the sorting icon size/spacing is not taken into account. 
![image](https://user-images.githubusercontent.com/3360588/67952122-c6d75380-fbba-11e9-860a-e046e139c592.png)

Additionally, when center aligned the content is not perfectly aligned with the label. The label may be smaller/larger than the cell contents. It seems continuing the pattern would direct us to not take the sorting icon into account when padding the right alignment. Thoughts?

JordanWSmith15 commented:
@tay1orjones @KarenGski @lukefirth Personally, I like aligning the text perfectly and ignoring the icon, but I'm fine with changing the design here, if it's easier to code it as Taylor is suggesting. We just need to make sure that we account for columns that are right-aligned and not sortable. In that case, we need to align the text. Here's an example: 
![Screen Shot 2019-10-31 at 10 16 22 AM](https://user-images.githubusercontent.com/21066152/67954539-8500da00-fbc7-11e9-8f20-be5a0d9ccecf.png)

davidicus commented on src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx:
```diff
@@ -419,6 +421,8 @@ const TableBodyRow = ({
             offset={offset}
             width={matchingColumnMeta && matchingColumnMeta.width}
+            align={align}
+            className={col.isSortable === true ? `data-table-${align}` : `data-table`}
```
we should use the classnames module to handle these.

davidicus replied:
Not sure why isSortable is tied with cell alignment. I have removed the tertiary and align defaults to `start`

davidicus commented:
Previously we seriously deviating from the Carbon defaults for padding. I have refactored the padding to use the Carbon base styles. Only padding added now is to resolve RTL issues.

sunilkotiyaibm commented:
> Previously we seriously deviating from the Carbon defaults for padding. I have refactored the padding to use the Carbon base styles. Only padding added now is to resolve RTL issues.



> Previously we seriously deviating from the Carbon defaults for padding. I have refactored the padding to use the Carbon base styles. Only padding added now is to resolve RTL issues.

Thanks David for refactoring the alignment code and adding the necessary changes.
In case of RTL, I added padding (td.data-table) to align table content with header.  Please have a look.

sunilkotiyaibm commented:
@davidicus @tay1orjones 
Please review updated changes in _data_table.scss for RTL case.

sunilkotiyaibm commented:
@davidicus @tay1orjones Please review the PR (574 table support column alignment
#626), fixed few alignment issue in _data_table.scss and updated storybook for alignment with few combination of Alignment and sorting.

tay1orjones commented:
:tada: This PR is included in version 2.13.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-addons-iot-react)
- [GitHub release](https://github.com/IBM/carbon-addons-iot-react/releases/tag/v2.13.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
